torch/lib/c10d/ProcessGroupNCCL.hpp

```diff
@@ -259,6 +249,15 @@ class ProcessGroupNCCL : public ProcessGroup {
           "FutureNCCL's value was internally set to NCCL collective's "
           "outputs or the return value of the callback.");
       value_ = std::move(value);
+
+      if (cudaEvents_ == nullptr) {
```

Comment:
Should this be an assert? Do we expect `cudaEvents_` be set twice on the same future object?

Reply:
After this PR, the second constructor of FutureNCCL leaves `cudaEvents_` unset. So this `if` here basically checks whether this instance was created with that second constructor. It is thus totally expected for this to happen, even without `markCompleted` being called twice.

I agree that it's not pretty, but let me assure you it's temporary. If you bear with me until #48506 you'll see that when we split FutureNCCL into CUDAFuture + a thin FutureNCCL subclass this will go away, because the two constructors will go one in each of these two classes, and thus these checks will automatically be replaced by inheritance.

Reply:
Oh, I get it now: when the future is created through the first constructor then `markCompleted` never gets called. Sorry I had missed that. In that case I think that, yes, this could become an assert. (It would have to be turned back into an `if` in #48505 though as there the first constructor will just call `markCompleted` internally).